daemon/src/listen_protocols.rs

```diff
@@ -0,0 +1,243 @@
+    identify::PROTOCOL,
+    xtra_libp2p_offer::PROTOCOL_NAME,
+);
+
+pub const REQUIRED_MAKER_LISTEN_PROTOCOLS: RequiredMakerListenProtocols =
+    RequiredMakerListenProtocols::new(
+        xtra_libp2p_ping::PROTOCOL_NAME,
+        identify::PROTOCOL,
+        rollover::v_2_0_0::PROTOCOL,
+        collab_settlement::PROTOCOL,
+    );
+
+/// Verify if the listen protocols that the `maker` supports are
+/// sufficient to fulfil the `requirements` of the taker.
+pub fn does_maker_satisfy_taker_needs(
+    maker: &HashSet<String>,
+    requirements: RequiredMakerListenProtocols,
+) -> Result<(), HashSet<String>> {
+    let missing_protocols = maker
+        .difference(&HashSet::<String>::from(requirements))
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    if !missing_protocols.is_empty() {
+        return Err(missing_protocols);
+    }
+
+    Ok(())
+}
+
+/// The set of protocols that the maker's `Endpoint` is listening for.
+#[derive(Clone, Copy)]
+pub struct MakerListenProtocols {
+    ping_v1: &'static str,
+    identify_v1: &'static str,
+    rollover_v1: &'static str,
+    rollover_v2: &'static str,
+    collaborative_settlement_v1: &'static str,
+}
+
+impl MakerListenProtocols {
+    pub const fn new(
+        ping_v1: &'static str,
+        identify_v1: &'static str,
+        rollover_v1: &'static str,
+        rollover_v2: &'static str,
+        collaborative_settlement_v1: &'static str,
+    ) -> Self {
+        Self {
+            ping_v1,
+            identify_v1,
+            rollover_v1,
+            rollover_v2,
+            collaborative_settlement_v1,
+        }
+    }
+
+    /// Construct a map of protocol identifiers to actor addresses.
+    ///
+    /// This is used so that the `Endpoint` knows who to delegate to
+    /// when receiving new inbound substreams.
+    pub fn inbound_substream_handlers<R1, R2>(
+        &self,
+        ping_v1: Address<pong::Actor>,
+        identify_v1: Address<identify::listener::Actor>,
+        rollover_v1: Address<
+            rollover::v_1_0_0::maker::Actor<command::Executor, oracle::AnnouncementsChannel, R1>,
+        >,
+        rollover_v2: Address<
+            rollover::v_2_0_0::maker::Actor<command::Executor, oracle::AnnouncementsChannel, R2>,
+        >,
+        collaborative_settlement_v1: Address<collab_settlement::maker::Actor>,
+    ) -> [(&'static str, MessageChannel<NewInboundSubstream, ()>); 5]
+    where
+        R1: rollover::v_1_0_0::protocol::GetRates + Send + Sync + Clone + 'static,
+        R2: rollover::v_2_0_0::protocol::GetRates + Send + Sync + Clone + 'static,
+    {
+        [
+            (self.ping_v1, ping_v1.into()),
+            (self.identify_v1, identify_v1.into()),
+            (self.rollover_v1, rollover_v1.into()),
+            (self.rollover_v2, rollover_v2.into()),
+            (
+                self.collaborative_settlement_v1,
+                collaborative_settlement_v1.into(),
+            ),
+        ]
+    }
```

Comment:
👀 These `inbound_substream_handlers` methods are the thing that is not ideal, but I didn't know how to make it better.

The problem I see is that there are no guarantees that we use all of the identifiers in `self` (or that we associate them with the correct `Address`, but that seems impossible).

Thinking about it now we could destructure `Self` and the compiler will tell us if we haven't used all of the fields. And it will complain if we add a new one! If we leave a nice comment it should be more or less obvious how the pattern should be used then.

Reply:
Nice idea! I applied it in the last `fixup` commit.
Additionally I applied the same pattern when converting into the `HashSet` for more assurance there.

I re-introduces the `NR_OF_SUPPORTED_PROTOCOLS` because it is another layer of checks we can apply and does not really add complexity.
Added a few more tests as well and fixed the `difference` function.

Would be great if you could have a look at the fixups :)